lib/retry.js

```diff
@@ -111,7 +129,16 @@ export default function retry(opts, task, callback) {
     function retryAttempt() {
         task(function(err) {
             if (err && attempt++ < options.times) {
-                setTimeout(retryAttempt, options.intervalFunc(attempt));
+                var proceed = true;
+                if(options.filter) {
+                    proceed = options.filter(err);
+                }
+
+                if(proceed) {
+                    setTimeout(retryAttempt, options.intervalFunc(attempt));
+                } else {
+                    callback.apply(null, arguments);
```

Comment:
`callback` is applied with `null` for context / this, and `arguments` for the actual arguments, so the error should be passed through. The new tests cover this. This is consistent with the scenario where we've exhausted all attempts and still had an error possibly.